### ISSUE TITLE
mhlo.scatter compilation failure

### ISSUE THREAD

**Issue by hawkinsp** [iree-org/iree#10774]
### What happened?

Originally reported as https://github.com/google/jax/issues/12786

```
Diagnostics:
/tmp/t.py:12:1: error: 'mhlo.scatter' op expects bounds of the scatter dimensions of updates to be same as the bounds of the corresponding dimensions of scatter indices. For scatter dimension 1, updates bound is 4 , scatter_indices bound is 1.
  return x.at[myslice].set(-99)
^
<unknown>:0: error: Failures have been detected while processing an MLIR pass pipeline
<unknown>:0: note: Pipeline failed while executing [`MHLOToMHLOPreprocessing` on 'func.func' operation: @main]: reproducer generated at `/tmp/ireeir/core-reproducer.mlir`
<unknown>:0: error: Failures have been detected while processing an MLIR pass pipeline
<unknown>:0: note: Pipeline failed while executing [`MHLOToMHLOPreprocessing` on 'func.func' operation: @main]: reproducer generated at `/tmp/ireeir/core-reproducer.mlir`
<unknown>:0: error: Failures have been detected while processing an MLIR pass pipeline
<unknown>:0: note: Pipeline failed while executing [`MHLOToMHLOPreprocessing` on 'func.func' operation: @main]: reproducer generated at `/tmp/ireeir/core-reproducer.mlir`
compilation failed
```

XLA accepts this IR happily, so I'd say it's an IREE bug.

### Steps to reproduce your issue

```
#loc0 = loc(unknown)
module @jit_setslice {
  func.func public @main(%arg0: tensor<8xi32> loc(unknown)) -> tensor<8xi32> {
    %0 = mhlo.constant dense<2> : tensor<i32> loc(#loc0)
    %1 = "mhlo.broadcast_in_dim"(%0) {broadcast_dimensions = dense<> : tensor<0xi64>} : (tensor<i32>) -> tensor<1xi32> loc(#loc1)
    %2 = mhlo.constant dense<-99> : tensor<i32> loc(#loc0)
    %3 = "mhlo.broadcast_in_dim"(%2) {broadcast_dimensions = dense<> : tensor<0xi64>} : (tensor<i32>) -> tensor<4xi32> loc(#loc2)
    %4 = "mhlo.scatter"(%arg0, %1, %3) ({
    ^bb0(%arg1: tensor<i32> loc(unknown), %arg2: tensor<i32> loc(unknown)):
      mhlo.return %arg2 : tensor<i32> loc(#loc3)
    }) {indices_are_sorted = true, scatter_dimension_numbers = #mhlo.scatter<update_window_dims = [0], scatter_dims_to_operand_dims = [0]>, unique_indices = true} : (tensor<8xi32>, tensor<1xi32>, tensor<4xi32>) -> tensor<8xi32> loc(#loc3)
    return %4 : tensor<8xi32> loc(#loc0)
  } loc(#loc0)
} loc(#loc0)
#loc1 = loc("jit(setslice)/jit(main)/broadcast_in_dim[shape=(1,) broadcast_dimensions=()]"("/tmp/t.py":12:1))
#loc2 = loc("jit(setslice)/jit(main)/broadcast_in_dim[shape=(4,) broadcast_dimensions=()]"("/tmp/t.py":12:1))
#loc3 = loc("jit(setslice)/jit(main)/scatter[update_consts=() dimension_numbers=ScatterDimensionNumbers(update_window_dims=(0,), inserted_window_dims=(), scatter_dims_to_operand_dims=(0,)) indices_are_sorted=True unique_indices=True mode=GatherScatterMode.FILL_OR_DROP]"("/tmp/t.py":12:1))

{-#
  external_resources: {
    mlir_reproducer: {
      pipeline: "stablehlo-legalize-to-hlo, func.func(mhlo-legalize-control-flow,iree-top-level-scf-to-cfg,iree-mhlo-to-mhlo-preprocessing{order-conv-features=true},canonicalize{  max-iterations=10 region-simplify=true top-down=true},shape-to-shape-lowering), convert-shape-to-std, func.func(canonicalize{  max-iterations=10 region-simplify=true top-down=true}), inline{default-pipeline= max-iterations=4 }, iree-util-demote-i64-to-i32, iree-util-demote-f64-to-f32, func.func(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,hlo-legalize-shape-computations,iree-mhlo-to-linalg-ext,iree-mhlo-to-linalg-on-tensors), reconcile-unrealized-casts, func.func(canonicalize{  max-iterations=10 region-simplify=true top-down=true}), iree-mhlo-verify-compiler-input-legality, iree-import-public, iree-import-ml-program, iree-sanitize-module-names, iree-abi-wrap-entry-points, inline{default-pipeline= max-iterations=4 }, func.func(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse), symbol-dce, iree-util-demote-f64-to-f32, func.func(iree-flow-convert-1x1-filter-conv2d-to-matmul,iree-flow-detach-elementwise-from-named-ops,iree-verify-input-legality),util.initializer(iree-flow-convert-1x1-filter-conv2d-to-matmul,iree-flow-detach-elementwise-from-named-ops,iree-verify-input-legality), linalg-named-op-conversion, iree-flow-expand-tensor-shapes, iree-util-fixed-point-iterator{max-iterations=10 pipeline=func.func(iree-util-simplify-global-accesses),util.initializer(iree-util-simplify-global-accesses),iree-util-apply-patterns,iree-util-fold-globals,func.func(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse),util.initializer(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse)}, iree-flow-tensor-pad-to-tensor-insert-slice{skip-one-linalg-use-case=false}, func.func(convert-elementwise-to-linalg,linalg-fold-unit-extent-dims{fold-one-trip-loops-only=false},iree-flow-interchange-generic-ops,resolve-shaped-type-result-dims,canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-flow-fusion-of-tensor-ops{fuse-multi-use=false multi-use-fusion-iteration=2},linalg-detensorize{aggressive-mode=false},canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-flow-split-reduction-ops,iree-flow-interchange-generic-ops,iree-flow-dispatch-linalg-on-tensors-pass{aggressive-fusion=false},iree-flow-capture-dispatch-dynamic-dims,canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse),util.initializer(convert-elementwise-to-linalg,linalg-fold-unit-extent-dims{fold-one-trip-loops-only=false},iree-flow-interchange-generic-ops,resolve-shaped-type-result-dims,canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-flow-fusion-of-tensor-ops{fuse-multi-use=false multi-use-fusion-iteration=2},linalg-detensorize{aggressive-mode=false},canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-flow-split-reduction-ops,iree-flow-interchange-generic-ops,iree-flow-dispatch-linalg-on-tensors-pass{aggressive-fusion=false},iree-flow-capture-dispatch-dynamic-dims,canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse), iree-flow-initialize-empty-tensors, iree-flow-outline-dispatch-regions, flow.executable(iree-util-strip-debug-ops),func.func(canonicalize{  max-iterations=10 region-simplify=true top-down=true}),util.initializer(canonicalize{  max-iterations=10 region-simplify=true top-down=true}), iree-flow-deduplicate-executables, flow.executable(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse),func.func(iree-flow-cleanup-tensor-shapes,canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse),util.initializer(iree-flow-cleanup-tensor-shapes,canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse), symbol-dce, iree-stream-verify-input, iree-stream-outline-constants, func.func(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-util-simplify-global-accesses),util.initializer(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-util-simplify-global-accesses), iree-util-apply-patterns, iree-util-fold-globals, iree-util-fuse-globals, iree-stream-conversion, iree-stream-verify-lowering-to-tensors, func.func(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-util-simplify-global-accesses),util.initializer(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-util-simplify-global-accesses), iree-util-apply-patterns, iree-util-fold-globals, iree-util-fuse-globals, iree-util-combine-initializers, func.func(iree-stream-encode-host-tensors),stream.executable(iree-stream-encode-device-tensors),util.initializer(iree-stream-encode-host-tensors), iree-stream-materialize-builtins, func.func(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-util-simplify-global-accesses),util.initializer(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-util-simplify-global-accesses), iree-util-apply-patterns, iree-util-fold-globals, iree-util-fuse-globals, func.func(iree-stream-materialize-copy-on-write),util.initializer(iree-stream-materialize-copy-on-write), iree-stream-elide-async-copies, iree-stream-refine-usage, func.func(iree-stream-schedule-execution,iree-stream-schedule-concurrency),util.initializer(iree-stream-schedule-execution,iree-stream-schedule-concurrency), iree-stream-propagate-timepoints, func.func(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-util-simplify-global-accesses),util.initializer(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-util-simplify-global-accesses), iree-util-apply-patterns, iree-util-fold-globals, iree-util-fuse-globals, iree-stream-verify-lowering-to-async, func.func(iree-stream-schedule-allocation,iree-stream-pack-constants,iree-stream-pack-allocations,iree-stream-layout-slices),util.initializer(iree-stream-schedule-allocation,iree-stream-pack-constants,iree-stream-pack-allocations,iree-stream-layout-slices), iree-util-propagate-subranges, func.func(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-util-simplify-global-accesses),util.initializer(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-util-simplify-global-accesses), iree-util-apply-patterns, iree-util-fold-globals, iree-util-fuse-globals, iree-stream-verify-lowering-to-cmd, func.func(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-util-simplify-global-accesses),util.initializer(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-util-simplify-global-accesses), iree-util-apply-patterns, iree-util-fold-globals, iree-util-fuse-globals, func.func(convert-scf-to-cf),util.initializer(convert-scf-to-cf), iree-util-fixed-point-iterator{max-iterations=10 pipeline=func.func(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-util-simplify-global-accesses),util.initializer(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-util-simplify-global-accesses),iree-util-apply-patterns,iree-util-fold-globals,iree-util-fuse-globals,iree-stream-elide-timepoints}, iree-stream-fuse-dispatch-bindings{alias-mutable-bindings=false}, iree-stream-pack-dispatch-operands, cse, iree-stream-fold-uniform-operands, iree-stream-annotate-dispatch-arguments, func.func(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-util-simplify-global-accesses),util.initializer(canonicalize{  max-iterations=10 region-simplify=true top-down=true},cse,iree-util-simplify-global-accesses), iree-util-apply-patterns, iree-util-fold-globals, iree-util-fuse-globals, symbol-dce, canonicalize{  max-iterations=10 region-simplify=true top-down=true}, cse, func.func(iree-util-simplify-global-accesses),util.initializer(iree-util-simplify-global-accesses), iree-util-apply-patterns, iree-util-fold-globals, iree-util-fuse-globals, iree-hal-assign-target-devices{targets=llvm-cpu}, iree-hal-verify-target-environment, iree-hal-materialize-interfaces, hal.executable(iree-hal-translate-executables), iree-hal-conversion, iree-hal-fixup-legacy-sync, canonicalize{  max-iterations=10 region-simplify=true top-down=true}, cse, func.func(iree-util-simplify-global-accesses),util.initializer(iree-util-simplify-global-accesses), iree-util-apply-patterns, iree-util-fold-globals, iree-util-fuse-globals, iree-hal-link-executables, iree-hal-resolve-export-ordinals, iree-hal-materialize-resource-caches, func.func(iree-hal-inline-device-switches),util.initializer(iree-hal-inline-device-switches), iree-hal-memoize-device-queries, canonicalize{  max-iterations=10 region-simplify=true top-down=true}, cse, func.func(iree-util-simplify-global-accesses),util.initializer(iree-util-simplify-global-accesses), iree-util-apply-patterns, iree-util-fold-globals, iree-util-fuse-globals, func.func(iree-hal-elide-redundant-commands),util.initializer(iree-hal-elide-redundant-commands), lower-affine, func.func(convert-scf-to-cf),util.initializer(convert-scf-to-cf), iree-util-combine-initializers, canonicalize{  max-iterations=10 region-simplify=true top-down=true}, cse, func.func(iree-util-simplify-global-accesses),util.initializer(iree-util-simplify-global-accesses), iree-util-apply-patterns, iree-util-fold-globals, iree-util-fuse-globals, hal.executable(iree-hal-serialize-executables), symbol-dce, inline{default-pipeline= max-iterations=4 }, symbol-dce, func.func(scf-for-loop-canonicalization,affine-loop-coalescing,loop-invariant-code-motion,convert-scf-to-cf,lower-affine,arith-unsigned-when-equivalent),util.initializer(scf-for-loop-canonicalization,loop-invariant-code-motion,convert-scf-to-cf,lower-affine,arith-unsigned-when-equivalent), iree-util-propagate-subranges, canonicalize{  max-iterations=10 region-simplify=true top-down=true}, cse, func.func(iree-util-simplify-global-accesses),util.initializer(iree-util-simplify-global-accesses), iree-util-apply-patterns, iree-util-fold-globals, iree-util-fuse-globals, iree-vm-conversion, vm.module(iree-vm-hoist-inlined-rodata,iree-vm-deduplicate-rodata), canonicalize{  max-iterations=10 region-simplify=true top-down=true}, cse, func.func(iree-util-simplify-global-accesses),util.initializer(iree-util-simplify-global-accesses), iree-util-apply-patterns, iree-util-fold-globals, iree-util-fuse-globals, vm.module(iree-vm-resolve-rodata-loads), inline{default-pipeline= max-iterations=4 }, symbol-dce, canonicalize{  max-iterations=10 region-simplify=true top-down=true}, cse, func.func(iree-util-simplify-global-accesses),util.initializer(iree-util-simplify-global-accesses), iree-util-apply-patterns, iree-util-fold-globals, iree-util-fuse-globals, vm.module(iree-vm-global-initialization), canonicalize{  max-iterations=10 region-simplify=true top-down=true}, cse, canonicalize{  max-iterations=10 region-simplify=true top-down=true}, vm.module(iree-vm-drop-empty-module-initializers), iree-util-drop-compiler-hints",
      disable_threading: false,
      verify_each: true
    }
  }
#-}
```

### Version information

iree-compiler 20221012.294 from PyPi.

**Comment by dcaballe** [comment]
We are also seeing a failure in our last [integrate](https://github.com/iree-org/iree/pull/10781) about this. We disabled `iree/compiler/InputConversion/MHLO/test/mhlo_to_mhlo_preprocessing.mlir.test` to move forward with the integration. That test should be re-enabled as part of a fix for this issue.

```
FAIL: IREE :: src/iree/compiler/InputConversion/MHLO/test/mhlo_to_mhlo_preprocessing.mlir (1 of 1)
******************** TEST 'IREE :: src/iree/compiler/InputConversion/MHLO/test/mhlo_to_mhlo_preprocessing.mlir' FAILED ********************
Script:
--
: 'RUN: at line 1';   iree-opt --split-input-file --verify-diagnostics --iree-mhlo-to-mhlo-preprocessing /usr/local/google/home/diegocaballero/iree3/compiler/src/iree/compiler/InputConversion/MHLO/test/mhlo_to_mhlo_preprocessing.mlir | FileCheck /usr/local/google/home/diegocaballero/iree3/compiler/src/iree/compiler/InputConversion/MHLO/test/mhlo_to_mhlo_preprocessing.mlir
--
Exit Code: 1

Command Output (stderr):
--
within split at /usr/local/google/home/diegocaballero/iree3/compiler/src/iree/compiler/InputConversion/MHLO/test/mhlo_to_mhlo_preprocessing.mlir:246 offset :4:8: error: unexpected error: 'mhlo.scatter' op expects scatter index leaf dimension to be within [0, rank(scatter_indices) + 1. rank(scatter_indices) is 0 and scatter index leaf dimension is 1.
  %0 = "mhlo.scatter"(%arg0, %arg1, %arg2) ({
       ^
within split at /usr/local/google/home/diegocaballero/iree3/compiler/src/iree/compiler/InputConversion/MHLO/test/mhlo_to_mhlo_preprocessing.mlir:262 offset :5:8: error: unexpected error: 'mhlo.scatter' op operand #2 must be tensor of integer or index values, but got 'tensor<7x11xf32>'
  %0 = "mhlo.scatter"(%arg0, %arg1, %arg2) ({
       ^
within split at /usr/local/google/home/diegocaballero/iree3/compiler/src/iree/compiler/InputConversion/MHLO/test/mhlo_to_mhlo_preprocessing.mlir:296 offset :5:8: error: unexpected error: Reduction-region must take 4 parameters, but takes 2 parameter(s)
  %0 = "mhlo.scatter"(%arg0, %arg1, %arg2) ( {
```

**Comment by hawkinsp** [comment]
Not fixed with iree-compiler 20221019.301

**Comment by astrotuna201** [comment]
Also not yet fixed in 20221020.302 (checked via originally reported  https://github.com/google/jax/issues/12786)
